### ISSUE TITLE
Script to autogenerates tests (like Snapshot testing)

### ISSUE THREAD

**Issue by integer256** [nalexn/ViewInspector#44]
Currently how [Snapshot](https://github.com/uber/ios-snapshot-test-case) testing works, is that you run the tests on a recording mode. Where you know that the views are correct, and later on, we can assert against that source of truth. Would it be possible to do something similar with this library?

I was wondering if would be helpful to create a script that auto-generates tests for a given View. 

I am imagining this script to be something like: 

1. Extract all the properties of a view, like color, frame, fonts, etc. 
2. Generate one test case for each property.
3. Probably some manual work is need it to customize these tests, by adding more, or being more specific, etc. 

The benefits will be that we can have a tool that combines the idea of Snapshot testing but using unit tests. 

Wdyt? Did you explore some idea like this before? 

Some of the inspiration for this idea are this [library of Android ](https://github.com/square/radiography) and this [layout inspector](https://fbflipper.com/docs/features/layout-plugin/).

**Comment by nalexn** [comment]
Hey @matiasvillaverde, there was a [proposal](https://github.com/nalexn/ViewInspector/issues/37) for improving the library in a way of making the lookup more dynamic and tolerant to the view structure changes.

Snapshotting or auto-generating the full path in the code saves time while you write the test, but the test remains fragile, where a tiny change in the view hierarchy could break it.

So far I have in plans digging towards that approach, but your proposal makes sense and I'll keep it in mind as a fall-back solution.

**Comment by matiasvillaverde** [comment]
Cool! It makes sense. In the case of snapshot, a change that breaks the tests can be a trigger to re-run the generation of the tests. 

Keep me in the loop, in case I can contribute to it, it is a topic that interest me.

**Comment by nalexn** [comment]
@integer256 with the recent release of the aforementioned dynamic query I can outline what is technically available now.

To begin with, each view in the hierarchy now can be located dynamically with `find` function, and after that, by calling `pathToRoot`, you can get a string representing the full inspection chain that leads to this view. For example:

```swift
let view = MyCustomView()
let sut = try view.inspect().find(ViewType.Text, where: { try $0.string == "Hello, world!" })
print(sut.pathToRoot)
// "anyView().hStack().group(1).vStack(0).text(0)"
```

The path can be used in place of `find`, and this thing resembles the process of "autogeneration" you were talking about. However, the usefulness of the replacement is questionable.

Looking at the functionality of the library you mentioned, I can add the feature of printing the entire view tree. It can, in theory, also print applied attributes and modifiers, however, it cannot yet calculate the layout frames.

Basically, ViewInspector can print a human-readable view structure with all the submitted values, and this dump will change every time something is tweaked inside the view. I cannot understand the use case where this can be helpful though - probably because I never used snapshot-generating libraries in tests. Let me know If such a full "dump" of the view hierarchy's structure is actually what you'd expect.

**Comment by integer256** [comment]
@nalexn That is great news! Thanks for the update.

I can imagine that would be helpful to have a method to print a human-readable view structure, as an intermediate step of writing the final test. Given that, due to the complexity of the hierarchy, sometimes it's difficult to inspect the correct view.

The new `find` function already simplifies the process of writing test a lot!
